torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp

```diff
@@ -0,0 +1,276 @@
+
+#include <torch/csrc/jit/codegen/cuda/lower_alias_memory.h>
+#include <torch/csrc/jit/codegen/cuda/expr_evaluator.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_printer.h>
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+//! Get string representation of Allocate size for symbolic comparison
+//!
+class SymbolicSizePrinter final : private OptOutConstDispatch {
+ public:
+  static std::string print_size(const kir::Allocate* alloc) {
+    SymbolicSizePrinter printer;
+    printer.handle(alloc->size());
+    return printer.os_.str();
+  }
+
+ private:
+  void handle(const Val* v) final {
+    OptOutConstDispatch::handle(v);
+  }
+
+  void handle(const Expr* e) final {
+    OptOutConstDispatch::handle(e);
+  }
+
+  void handle(const kir::Int* node) final {
+    if (auto def = FusionGuard::getCurFusion()->origin(node)) {
+      os_ << "( ";
+      handle(def);
+      os_ << " )";
+      return;
+    } else if (node->isSymbolic()) {
+      os_ << "i" << node->name();
+    } else {
+      os_ << *node->value();
+    }
+  }
+
+  void handle(const kir::NamedScalar* node) final {
+    os_ << node->name();
+  }
+
+  void handle(const kir::BinaryOp* node) final {
+    if (auto inline_bop = inline_op_str(node->getBinaryOpType())) {
+      handle(node->lhs());
+      os_ << " " << inline_bop.value() << " ";
+      handle(node->rhs());
+    } else {
+      os_ << node->getBinaryOpType() << "(";
+      handle(node->lhs());
+      os_ << ", ";
+      handle(node->rhs());
+      os_ << ")";
+    }
+  }
+
+ private:
+  std::stringstream os_;
+};
+
+//! Scan through Kernel IR to reuse Allocation nodes
+//! via pointer aliasing
```

Comment:
can you please elaborate with an overview of the algorithm? (how do we track buffer lifetimes, ...)

Reply:
:+1: